File: leetcode/python/150.evaluate-reverse-polish-notation.py
from typing import List


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        op = {'+', '-', '*', '/'}
        for i, v in enumerate(tokens):
            if v in op:
                n = stack[-1]
                stack.pop()
                m = stack[-1]
                stack.pop()

                def evalVal(op, m, n):
                    if op == '+':
                        return m + n
                    elif op == '-':
                        return m - n
                    elif op == '*':
                        return m * n
                    else:
                        return int(m / n)
                stack.append(evalVal(v, m, n))
            else:
                stack.append(int(v))
        return stack[0]

File: leetcode/cpp/150.evaluate-reverse-polish-notation.cpp
#include "oj_header.h"

class Solution
{
public:
    int evalRPN(vector<string> &tokens)
    {
        unordered_set<string> set{"+", "-", "*", "/"};
        stack<int> _stack;
        for (size_t i = 0; i < tokens.size(); i++)
        {
            if (set.find(tokens[i]) == set.end())
            {
                _stack.push(stoi(tokens[i], nullptr));
            }
            else
            {
                int m = _stack.top();
                _stack.pop();

                int n = _stack.top();
                _stack.pop();
                _stack.push(eval(tokens[i], n, m));
            }
        }

        return _stack.top();
    }

    int eval(const string &op, int m, int n)
    {
        if (op == "+")
            return m + n;

        if (op == "-")
            return m - n;
        if (op == "*")
            return m * n;
            
        return m / n;
    }
};

